packages/enzyme-adapter-utils/src/Utils.js

```diff
@@ -92,24 +92,55 @@ export function nodeTypeFromType(type) {
   return 'function';
 }
 
-function isIterable(obj) {
-  return (
-    obj != null &&
-    typeof Symbol === 'function' &&
-    typeof Symbol.iterator === 'symbol' &&
-    typeof obj[Symbol.iterator] === 'function'
+function getIteratorFn(obj) {
+  const iteratorFn = obj && (
+    (
+      typeof Symbol === 'function' &&
+      typeof Symbol.iterator === 'symbol' &&
+      obj[Symbol.iterator]
+    ) ||
+    obj['@@iterator']
   );
+
+  if (typeof iteratorFn === 'function') {
+    return iteratorFn;
+  }
+
+  return undefined;
+}
+
+function isIterable(obj) {
+  return !!getIteratorFn(obj);
 }
 
 export function isArrayLike(obj) {
-  return Array.isArray(obj) || (isIterable(obj) && typeof obj !== 'string');
+  return Array.isArray(obj) || (typeof obj !== 'string' && isIterable(obj));
 }
 
 export function flatten(arrs) {
-  return arrs.reduce(
-    (flattened, item) => flattened.concat(isArrayLike(item) ? flatten([...item]) : item),
-    [],
-  );
+  let flatArrs = [];
+
+  const iteratorFn = getIteratorFn(arrs);
+  const iterator = iteratorFn.call(arrs);
+
+  let step = iterator.next();
+
+  while (!step.done) {
+    const item = step.value;
+    let flatItem;
+
+    if (isArrayLike(item)) {
+      flatItem = flatten(item);
+    } else {
+      flatItem = item;
+    }
+
+    flatArrs = flatArrs.concat(flatItem);
+
+    step = iterator.next();
+  }
+
+  return flatArrs;
```

Comment:
This feels like a lot of added complexity when `flatten` should almost always be called with arrays.  It just seems like we're doing the same work that `[...iterable]` is doing, except that this approach won't benefit from any internal optimizations that the spread operator may have.

Reply:
Our use case is running our unit test suite on PhantomJS, where some of our components return Immutable List instance of React elements. As described in https://github.com/airbnb/enzyme/issues/1328, that feature was recently added but was not working on PhantomJS due the lack of support for Symbol.iterator. I agree the increased complexity is evident, and that's why I aimed for good test coverage. The main reason to make this change is to catch up with React, which is very flexible with the children is able to handle, and switching from v2 to v3 may break some middle to large projects using a pattern like us and results in additional effort to make the bump. Really appreciate your comments.

Reply:
@aweary, what if I add below code at the beginning of `flatten` function?

```
if (Array.isArray(arrs)) {
  return arrs.reduce(
    (flattened, item) => flattened.concat(isArrayLike(item) ? flatten(item) : item),
    [],
  );
}
```

That way `flatten` will be as performant as it used to before my change, for the most common case where `arrs` is a regular JS `Array`. But gracefully fallbacks to support any kind of iterable children React can also handle.

Let me know your thoughts. cc @ljharb

Reply:
Seems reasonable to me.

Reply:
Cool, I'm going to make that change and push.